xmake/rules/platform/windows/idl/idl.lua

```diff
@@ -0,0 +1,177 @@
+        if fileconfig.undefines then
+            table.join2(undefs, fileconfig.undefines)
+        end
+    end
+
+
+    local flags = {"/nologo"}
+    table.join2(flags, table.wrap(target:values("idl.flags")))
+
+    -- specify warn levels
+    local warns = target:get("warnings")
+    if warns == "none" then
+        table.insert(flags, "/W0")
+        table.insert(flags, "/no_warn")
+    elseif warns == "less" then
+        table.insert(flags, "/W1")
+    elseif warns == "more" then
+        table.insert(flags, "/W2")
+    elseif warns == "extra" then
+        table.insert(flags, "/W3")
+    elseif warns == "error" then
+        table.insert(flags, "/WX")
+        table.insert(flags, "/W4")
+    end
+
+    -- add include dirs, defines and undefines from compiler flags
+    for _, inc in ipairs(incs) do
+        table.insert(flags, "/I")
+        table.insert(flags, path(inc))
+    end
+
+    for _, def in ipairs(defs) do
+        table.insert(flags, "/D")
+        table.insert(flags, def)
+    end
+
+    for _, undef in ipairs(undefs) do
+        table.insert(flags, "/U")
+        table.insert(flags, undef)
+    end
+
+    table.join2(flags, {
+        "/out",    path(autogendir),
+        "/header", name .. ".h",
+        "/iid",    name .. "_i.c",
+        "/proxy",  name .. "_p.c",
+        "/tlb",    name .. ".tlb",
+        "/cstub",  name .. "_c.c",
+        "/sstub",  name .. "_s.c",
+        "/server", (enable_server and "stub" or "none"),
+        "/client", (enable_client and "stub" or "none"),
+        path(sourcefile)
+    })
+
+    depend.on_changed(function() 
+        progress.show(opt.progress or 0, "${color.build.object}generating.idl %s", sourcefile)
+        os.vrunv(midl.program, flags, { envs = msvc:runenvs() })
+    end, {files = sourcefile, 
+         dependfile = path.join(autogendir, path.basename(sourcefile) .. ".idl.d") }
+    )
+end
+
+-- add *.idl for rc file
+
+function configure(target)
+    local sourcebatch = target:sourcebatches()["platform.windows.idl"]
+    if sourcebatch then
+        local autogendir = path.join(target:autogendir(), "platform/windows/idl")
+        os.mkdir(autogendir)
+        target:add("includedirs", autogendir, {public = true})
+    end
+end
+
+function build_idlfiles(target, jobgraph, sourcebatch, opt)
+    local mysources = {}
+    local autogendir = path.join(target:autogendir(), "platform/windows/idl")
+
+    local addsrc = function (sourcename, suffix)
+        local fullfile = path.join(autogendir, sourcename .. suffix)
+        if os.exists(fullfile) then
+            table.insert(mysources, fullfile)
+        end
+    end
+
+    for _, sourcefile in ipairs(sourcebatch.sourcefiles) do
+        local fileconfig = target:fileconfig(sourcefile)
+        local enable_proxy = true
+        if fileconfig then
+            if fileconfig.proxy ~= nil then
+                enable_proxy = fileconfig.proxy
+            end
+        end
+
+        generate_single(target, sourcefile, opt)
```

Comment:
we should add call `generate_single` in jobgraph.

like this https://github.com/xmake-io/xmake/blob/52950c684c10690a9db791e4919760316cce26e6/xmake/rules/c%2B%2B/modules/modules_support/builder.lua#L302

then use `jobgraph:add_orders` and group to add run deps

https://github.com/xmake-io/xmake/blob/52950c684c10690a9db791e4919760316cce26e6/xmake/rules/c%2B%2B/modules/modules_support/builder.lua#L454